File: patterns/structural/decorator/decorator.ts
abstract class Beverage {
  protected description: string;

  getDescription(): string {
    return this.description;
  }

  abstract cost(): number;
}

abstract class ConfimentDecorator extends Beverage {
  protected beverage: Beverage;
  abstract getDescription(): string;
}

class Expresso extends Beverage {
  constructor() {
    super();
    this.description = "Expresso";
  }

  cost(): number {
    return (1.99);
  }
}

class DarkRoast extends Beverage {
  constructor() {
    super();
    this.description = "DarkRoast";
  }

  cost(): number {
    return (1.12);
  }
}

class HouseBlend extends Beverage {
  constructor() {
    super();
    this.description = "HouseBlend";
  }

  cost(): number {
    return (0.89);
  }
}

class Mocha extends ConfimentDecorator {
  
  constructor(beverage: Beverage) {
    super();
    this.beverage = beverage;
  }

  getDescription(): string {
    return this.beverage.getDescription + ", Mocha";
  }

  cost(): number {
    return this.beverage.cost() + 0.20;
  }
}

class Soy extends ConfimentDecorator {
  
  constructor(beverage: Beverage) {
    super();
    this.beverage = beverage;
  }

  getDescription(): string {
    return this.beverage.getDescription + ", Soy";
  }

  cost(): number {
    return this.beverage.cost() + 0.12;
  }
}

class Whip extends ConfimentDecorator {
  
  constructor(beverage: Beverage) {
    super();
    this.beverage = beverage;
  }

  getDescription(): string {
    return this.beverage.getDescription + ", Whip";
  }

  cost(): number {
    return this.beverage.cost() + 0.12;
  }
}

const beverage = new Expresso();
console.log('1) ', beverage.getDescription(), ' ', beverage.cost());

let beverage2 = new DarkRoast();
beverage2 = new Mocha(beverage2);
beverage2 = new Soy(beverage2);
beverage2 = new Whip(beverage2);

console.log('1) ', beverage2.getDescription(), ' ', beverage2.cost());

File: patterns/structural/decorator/decorator.cpp
#include <iostream>
#include <string>

class Beverage {
protected:
    std::string _description;

public:
    virtual std::string getDescription() {
        return (_description);
    }

    virtual double cost(void) = 0;
};

class CondimentDecorator : public Beverage {
public:
    virtual std::string getDescription(void) = 0;
};

class Expresso : public Beverage {
public:
    Expresso() {
        _description = "Expresso";
    }

    double cost(void) {
        return (1.99);
    }
};

class DarkRoast : public Beverage {
public:
    DarkRoast() {
        _description = "Dark roast";
    }

    double cost(void) {
        return (1.12);
    }
};

class HouseBlend : public Beverage {
public:
    HouseBlend() {
        _description = "House Blend Coffe";
    }

    double cost() {
        return (0.89);
    }
};

class Mocha : public CondimentDecorator {
private:
    Beverage *_beverage;

public:
    Mocha(Beverage *beverage) {
        _beverage = beverage;
    }

    ~Mocha() {
        free(_beverage);
    }

    std::string getDescription(void) {
        return (_beverage->getDescription() + ", Mocha");
    }

    double cost() {
        return (_beverage->cost() + 0.20);
    }
};

class Soy : public CondimentDecorator {
private:
    Beverage *_beverage;

public:
    Soy(Beverage *beverage) {
        _beverage = beverage;
    }

    ~Soy() {
        free(_beverage);
    }

    std::string getDescription(void) {
        return (_beverage->getDescription() + ", Soy");
    }

    double cost() {
        return (_beverage->cost() + 0.60);
    }
};

class Whip : public CondimentDecorator {
private:
    Beverage *_beverage;

public:
    Whip(Beverage *beverage) {
        _beverage = beverage;
    }

    ~Whip() {
        free(_beverage);
    }

    std::string getDescription(void) {
        return (_beverage->getDescription() + ", Whip");
    }

    double cost() {
        return (_beverage->cost() + 0.04);
    }
};

int main(void) {
    Beverage *beverage = new Expresso();
    std::cout << "1) " << beverage->getDescription() << " $" << beverage->cost() << "\n";

    Beverage *beverage2 = new DarkRoast();
    beverage2 = new Mocha(beverage2);
    beverage2 = new Mocha(beverage2);
    beverage2 = new Whip(beverage2);
    std::cout << "2) " << beverage2->getDescription() << " $" << beverage2->cost() << "\n";

    Beverage *beverage3 = new HouseBlend();
    beverage3 = new Soy(beverage3);
    beverage3 = new Mocha(beverage3);
    beverage3 = new Whip(beverage3);
    std::cout << "3) " << beverage3->getDescription() << " $" << beverage3->cost() << "\n";

    free(beverage);
    free(beverage2);
    free(beverage3);
    return (0);
}